### DISCUSSION TITLE
Lower core web vitals esp. TTI and TBT

### DISCUSSION THREAD

**Discussion by devhiko** [vercel/next.js#40107] [Help] [answered]
Hey, i have an application based on nextjs and styled with bootstrap and boosted with CMS as Contentful. As altough i think an optimized component code and well structured codebase i have, even so i have too high (according to lighthouse report and my sigths) my report values is to high like TBT: ^900ms,~ and TTI: ^5s~ and i giving error and suggestions about decreasing this, but none of them worked for me to decrease this and get rid of them. I have read somewhere at web like this: **Next.js apps have fast layout and user experince but lower TTI and TBT (render blocking -> white page with almost nothing w/out spinner for loading and lasts long time in contrast of other apps like structured with react or vanilla js.** And even more, i have used typescript and react-bootstrap framework for styling and other pages. My code is using imports dynamic and optimizes images with <Image> tag. Most of time i have searched on the net for this optimizing Core Web Vitals thing and writing optimized code. But none of them relieved this pain of suffering the main thread. I giving also error, **Minimize main thread work, Decrease javascript execution time** As a question for need help as mainly, IS THIS ISSUE ABOUT NEXTJS CORE, OR MY CODEBASE OR LİGHTHOUSE REPORT STABILTY ISSUES, i read an article about lighthouse audits's stability and reliability. Please help me for further, because i am newbie about at this environment but i want to develop faster and optimized web apps. In an application the reports gives many errors about **overwhelmed and getting hotter core**, at a scale even bigger application what will be worse more than i think. Here is a link of my question for detail: [stackoverflow question](https://stackoverflow.com/questions/73547632/lower-core-web-vitals-esp-tti-and-tbt)

**Comment by icyJoseph** [comment]
Got a repo to look at? Are you using lighthouse as a CLI?

There's a concept of CPU power, when I run lighthouse on my machine, I get like 27xx CPU power on 4x slowdown... https://github.com/GoogleChrome/lighthouse/blob/master/docs/throttling.md#cpu-throttling

It is worth investigating about a property slowdown multiplier for your machine and running the CLI tool to get stable results.

In my case since I have 27xx, the page https://lighthouse-cpu-throttling-calculator.vercel.app/ recommends, to use x9 slowdown multiplier, and with that I get sorta regular results. 

I'd say, if there's no actual issues for your users, then you are probably alright. Fine tune the slowdown and use the CLI to get reliable results.

**Comment by devhiko** [comment]
Yeah, i have that. Here is the link: [repo](https://github.com/devhiko/dialer) and thanks for help. No i am not using it in CLI, normal with devtools and @6x slowdown, slow 3g, **but suprisingly it didn't change at any throttling levels even no throttling cpu and network.** I used it with recommend from mobile device constraints.  I did it and i give a number like : **"benchmarkIndex": 1069** and at [score](https://lighthouse-cpu-throttling-calculator.vercel.app/) -> 2.5x gave. In my case i better to use with no throttling for accurate results. but even i said before there is no difference between results with throttling and w/out. And with recommendation as 2.5x i tried again.
And i would like to say, my code is running too slow and render blocking as you see. Please check my repo and maybe you help for further. And here is for comparison: 
![a](https://user-images.githubusercontent.com/99843102/187664184-ef67645a-c7b9-4b6d-ad1e-ffe0c719bba4.png)
at left @2.5x CPU slowdown and at right no throttling.

**Reply by icyJoseph** [reply]
The reason to use the CLI is that, your browser session is kind of polluted. Often people use incognito mode to run lighthouse, could you at least try with that? Or is it already on incognito mode?

**Reply by icyJoseph** [reply]
Btw, lighthouse does its own throttling, AFAIK,  it doesn't apply your CPU or Network slowdown, it applies its own. 

```
lighthouse "http://localhost:3000" --view  --throttling.cpuSlowdownMultiplier=2.5
```

This is why the CLI is a much more reliable tool.

**Reply by icyJoseph** [reply]
Last but not least, I'd rather see a deployed version of this app, not just the repo, because there's a few credentials I'd need to setup in order to spin it up.

**Reply by devhiko** [reply]
Yes i always tried with at incognito and no extensions active, but lightouse CLI spinned up in a blank normal window and i can't change it :) Yeah the report at left is ran with command above: lighthouse "http://localhost:3000" --view  --throttling.cpuSlowdownMultiplier=2.5 and it did not change as you see at right,. Yes you are right and i am gonna deploy it to share with you and here is the link: [product deploy](https://dialer-two.vercel.app/son)

And maybe you suggest sth with my code to optimize for better results or maybe it is normal for app that scale. And i dont know why, the style broken :(, maybe it cause from react-bootstrap package but for performance testing is enough for research :), Thanks for help, again.

**Reply by icyJoseph** [reply]
Hehe I got between 95 and 100 consistently with x9 slowdown...

However, this [result](https://googlechrome.github.io/lighthouse/treemap/?gzip=1#H4sIAAAAAAAAE62WS2/jNhCA/4qhsyVz+GaO7QLby6It0pyKIuBj6LCRJVWikzUC//fSTk6Ksgiq+mYOZ77RPPlStQ9jdfNSjfjPEaeM4W5sq5vqIedhutntQrItjnV+7psnHD22jR2G3dR31baKqbPt56/bY0h5urAmP6Yh13lEPNihDjbby3EKxdKSbFvllFss0turdPPHq3Tz5VUa8FUrFc5NdTdh2MR+3LzZ2BQXyqXJ92Mx0R3b9u3PlzQNrT1968PFdOqKzsHm9IRXk9mm9uptPg0X+cyjrmgV8Z8vVWcP+MkQjDj1x3L20ylflIEqed5+0sJ9h9/zbsrFQ7/zD8fucdoNfXuKqW2n2ktllQjgZDQmeN38Pb3nkVUw2QjrnXIoLWhqvfKLEKmo2FbH7ljS8HbEuBCr0JpBowwP2mjqnLPIPFuEa0lmbKX4KrTgtKGRGap18Ki9MJwvopWhM/S6cPMAzOkoGgAuOAnOBhaVsctR50RQDvO4EyZgXYVpwhvNvdKea0EQKBK76AAQY4yaB58BsHXhF6a2mkqiS8E5Hq0Wy5kHxYyZ4YEAU+vwumGMaDQECPMGtAjLdKr5vOqN0etiT41qoouoIlEUZQAdxHLyAfT7T4dV8JLPRitpkEtCaUBh4nLdU0XkDM6ZWcVWwJtAvHPKoGTKBArwQdERyWdwSWFdzp/RDdY/1ky6MlQZBgNKErVcdoLRedtTJtY5EMei+dyPj7VxIkj0vPy8JvGDzuMgyLz0OTF03fw52NTVlmsEwzgNElBr+UEWJCPz2SOE5Otaf7B7nHb3RVyDUJwzTtCWRpNkeQSCpO+8YGV8r1t6TNUs2EjBUm6o90Qt9yCA4PM+KGN7XSW8hqC8HWrFXaS2bN+ygcoeIh9EQM8n8H8tgl9ixN+ffv51/7UG3g6Ph6+/fT/t7t0xteGb7VIsb8Xl9S/mOfh/PZim/Q/5APNWIOe/zufztvJ9F9P+FnNO3f76smt7b6/PSuzqu9uq3PoXXgh17w8LAAA=), seems to suggest that there's quite a lot of unused JavaScript in chunk `/_next/static/chunks/4d13b8f5.114540dbad3f79ac.js`.

It might not be related but you do:

```js
import { Spinner } from "react-bootstrap";
``` 

When it might be better to do, or which ever applies for the Spinner:

```js
import { Spinner } from 'react-bootstrap/lib/Spinner'; 
```

Refer to https://stackoverflow.com/questions/53306947/why-is-it-not-ideal-to-import-react-bootstrap-using-destructuring, for a slightly expanded.

I'd recommend using [`@next/bundle-analyzer`](https://www.npmjs.com/package/@next/bundle-analyzer), to get a visual help into which libraries are packaged whole, when perhaps you just need slight sections of them.

**Comment by devhiko** [comment]
Yeah, it related with Contentful library files with axios. i cant handle that :) `/_next/static/chunks/4d13b8f5.114540dbad3f79ac.js.` And for your system, it seems to be will Mac, and i have a oldschool guy with 6 yrs. I expect this app runs at constrained mobile devices, it will enough for me. I know which libs bloated, but i have to use that for styling and data fetching like CMS, React-Bootstrap, typescript and so on. Maybe you have suggestions about for my codebase like optimizing components or page speed you know. And btw, can you share a lighthouse report at your own machine i am gonna look for TTI, TBT, and more. :) Thanks for all, again. Maybe core vercel or lighthouse developers will answer our discussion and give some advices about better PSI. And for last, with tree-shaking there is no difference about {} from /lib or /. here -> , but for wonder i am gonna try that :D But i dont know how can i use with my next.config with next/bundle-analyzer it structured different from there as default, maybe you can help.

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  env: {
    C_SPC_ID: **************
    C_ACC_TKN: ***************
    C_MNG_TKN: *****************
  },
  i18n: {
    locales: ["tr"],
    defaultLocale: "tr",
  },
};

module.exports = nextConfig;
```

**Reply by icyJoseph** [reply]
> Yeah, it related with Contentful library files with axios. i cant handle that :) /_next/static/chunks/4d13b8f5.114540dbad3f79ac.js. And for your system, it seems to be will Mac, and i have a oldschool guy with 6 yrs. I expect this app runs at constrained mobile devices, it will enough for me.

Yeah, but the x9 slowdown should help us mimic a slow device. I'd say it is almost better to try this out on a device.

> I know which libs bloated, but i have to use that for styling and data fetching like CMS, React-Bootstrap, typescript and so on. Maybe you have suggestions about for my codebase like optimizing components or page speed you know. 

Well, one thing I am trying to understand is why the client bundle is packing contentful stuff. Probably some imports are mixed up, but from looking at: https://github.com/devhiko/dialer/blob/master/pages/son/index.tsx, if we pass it through https://next-code-elimination.vercel.app/, we see that it eliminates contentful from the bundle, so in principle there's no reason why it should be present.

We must look at who imports from what, because most likely, a util function, or a type being imported incorrectly is telling webpack to bundle the library to the client...

I'll read the repo a bit more to try and figure it out.

**Reply by icyJoseph** [reply]
For the bundle analyzer do:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  env: {
    C_SPC_ID: "**************",
    C_ACC_TKN: "***************",
    C_MNG_TKN: "*****************",
  },
  i18n: {
    locales: ["tr"],
    defaultLocale: "tr",
  },
};

const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})

module.exports = withBundleAnalyzer(nextConfig);
```

Then put this in your package.json:

```
"scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "test": "jest",
    "stats": "ANALYZE=true next build"
  },
```

And run it with `yarn stats` or `npm run stats`.

**Reply by devhiko** [reply]
Thanks, i am gonna try a version no-tree-shaked with default imports and bundle analyzer. But i think it required for fetching data from Contentful client where setted up at /cms/setup yo see. And so, maybe there is a way to fix that or it really needs for this line for working properly. But also i dont know a way importing client from contentful the library seems not suitable for tree-shaking. And further, you can play with my code as you want :D,

**Comment by icyJoseph** [comment]
@devhiko would you please do this change to your types?

```ts
import type { Entry, EntryCollection } from "contentful";
import type { MouseEvent } from "react";
import type { OffcanvasProps } from "react-bootstrap";
```

And then try again?

I suspect the way you currently have it, is loading the entire library, and not just the types. In general when possible, just import the type via `import type` syntax, https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html.

**Reply by devhiko** [reply]
Ok, i am gonna try that also, but i knew it before import type is deprecated for typescript you know. import ... is approved for it. ..aand i did that, please check out :)

**Reply by icyJoseph** [reply]
https://www.typescriptlang.org/docs/handbook/modules.html#importing-types ? what do you mean deprecated? Have you got some info regarding that?

Please do deploy your changes, I cannot build the project without the CMS keys, and I am already running short on time for the day :)

**Reply by devhiko** [reply]
Prior to TypeScript 3.8, you can import a type using import. With TypeScript 3.8, you can import a type using the import statement, i supposed that it works properly :D but not maybe. And meanwhile i redeployed it at vercel. you can check that out

**Reply by devhiko** [reply]
and for further investigation i can share my tokens with you, that is not a problem with me. Have it your way your own.

**Reply by icyJoseph** [reply]
It didn't help so much because of components like: https://github.com/devhiko/dialer/blob/1623db837f690f316e60f33e950da8a81cdf636a/components/kisiler/KisiSilModal.tsx

which import from `cms/setup` and that invokes the entire `contentful-management` library.

I guess the only other option would be move the CRUD layer to API routes, and via `fetch` POST do these operations, then the client won't need to use contentful packages at all... 

Sort of like `fetch("/api/cms/kisiEkle", {method: 'POST', body: {})`, and that way you edit the content, then in `pages/api/cms/kisiEkle` you invoke the contentful API, safely and cheaply, because it runs on the server.

**Comment by devhiko** [comment]
Hmm, it sounds magical i am enlightened. I dont know how to  structure my /api routes, i did not use it all and for honesty i did not need it. but i can learn and try it. and i am giving congrats for your knowledge and effort i dont know so much to you. I heard about react server components and i did not figure it out how and i developing client side only, it maybe relative with node.js and so i did not used till now. Maybe it helps for that approach you know

**Reply by icyJoseph** [reply]
Server Components as a technology is experimental right now, I think you can pull this off with API routes.

**Comment by devhiko** [comment]
I am gonna try all the suggestions and approaches and will refactor, optimize and deploy my code and after run the lighthouse CLI commands again, i hope it will be better at all devices. Dude thanks for conversation and effort, i learned lot from you, and if you need help please @mention me about.

**Reply by icyJoseph** [reply]
No worries, if you struggle with the api routes stuff, we can try to setup a meets meeting or something similar to hack it together. 

API routes has to be the breakthrough here.

**Reply by devhiko** [reply]
I have a confidence i can figure that out :D but if not i ll be there for hacking with you :) . i can not speak english well but with chat messages it ll be better, please tell me when if will there is a timezone difference you know.

**Comment by devhiko** [comment, marked as answer]
I tried all optimizations and suggestions except /api stuff. and gotcha i hit the 100 at all. here is my result: 
![a](https://user-images.githubusercontent.com/99843102/187786062-5ae83609-87eb-49b0-b5d4-4848c08f9262.png)
But although, i have not tried /api stuff and wondering about that. Maybe can you give little help with that. There is a no requirement for that maybe, but i better to know it. Thanks for all journey, have a nice work !

**Reply by icyJoseph** [reply]
Sweet! Good job! now the problem is to stay near the 100 as the application scales! 😬 🎉

**Comment by devhiko** [comment]
Ya, but i finished it already, i am not sure about i am gonna make an app
bigger than this, as clear, i developed that on my own but for further or
bigger, i will be not enough i think.

1 Eyl 2022 Per 10:32 tarihinde Joseph ***@***.***> şunu yazdı:

**Reply by incora-dev** [reply]
@devhiko hello do you have any hacks to improve performance, i have similar issue for performance of next js app in page speed

**Reply by devhiko** [reply]
@incora-dev Hey, i would like to help you for sure. Maybe we can connect in some way like discord cuz its a deep topic and depend on many things.

**Reply by incora-dev** [reply]
@devhiko give me your nickname, for sure let's connect

**Reply by devhiko** [reply]
Yea, its `@oldcoder` at discord, reach me out i am waiting for you.

**Reply by incora-dev** [reply]
@devhiko i've sended you request in discord

**Comment by 11koukou** [comment]
Hello devhiko,

I have managed to maintain a constant score around 95 for my website's homepage which is a fully functional e-commerce fashion site with many images and components like carousel etc, and google analytics/facebook pixel external scripts too! I will post my own experience from pagespeed after more than 2 months of experimenting:

First of all, since pagespeed/ligthouse test the mobile viewport by default, make sure to optimize your site for the mobile viewport first. That means to control your components and dom size and most important, your image sizes. Also keep in mind that you need to minimize the rendering work at the browser which means that you need to pre-render as much of html as possible using static generation. Use the react profiler plugin for chrome devtools and make sure that you do not have any unnecessary re renders in page and other heavy components. 
Moreover in case that you are using external scripts try moving them from main thread to service workers using partytown. It takes some time but after very careful work you can manage to have a lightning fast mid to large scale website! Don't hesitate to ask me anything you need. I would post my website's url but since it still uses non copyrighted images and non production api tokens for external services I can not publish it yet..

Thank you

**Reply by devhiko** [reply]
Hey, thanks for your insights. But actually i have a problem that not solved and there is no way it seem to solve. I am using default postcss config for purging css according to lighthouse suggestion but **i am losing my styles after deploy or even production build.** When i delete the config file as postcss.config you know, it reversed back to original, everything is good but my scores goes down to start status. Maybe you can check my repo and suggest some little things. And maybe you heard, **we talked about decreasing client bundle size with moving CRUD layer to server side**, i dont know it helps but i have no hope to correct my values near to 100 or 90 at least. Because **i have a small app i think, at scales it broken down. At now even at x6 slow cpu and slow 3g throttling, it is nearly freezing.** Please help me :( @11koukou
